python/ml4ir/model/losses/listwise_losses.py

```diff
@@ -8,4 +12,40 @@ def _make_loss_fn(self, **kwargs):
         Additionally can pass in record positions to handle positional bias
 
         """
-        raise NotImplementedError
+        bce = losses.BinaryCrossentropy(reduction=Reduction.SUM)
+        mask = kwargs.get("mask")
+        # cce = losses.CategoricalCrossentropy(reduction=Reduction.SUM_OVER_BATCH_SIZE)
+
+        def _loss_fn(y_true, y_pred):
+            # Mask the padded records
+            y_true = tf.gather_nd(y_true, tf.where(tf.equal(mask, tf.constant(1.0))))
+            y_pred = tf.gather_nd(y_pred, tf.where(tf.equal(mask, tf.constant(1.0))))
+
+            # Reshape the tensors
+            y_true = tf.expand_dims(tf.squeeze(y_true), axis=-1)
+            y_pred = tf.expand_dims(tf.squeeze(y_pred), axis=-1)
+
+            return bce(y_true, y_pred)
+            # return cce(y_true, y_pred)
+
+        return _loss_fn
+
+    def _final_activation_op(self):
+        # Without masking padded records
+        softmax = layers.Softmax(axis=-1, name="ranking_scores")
+
+        # Listwise Top 1 RankNet Loss
+        def masked_softmax(logits, mask):
+            """
+            NOTE:
+            Tried to manually compute softmax with tf operations,
```

Comment:
why would you manually compute this with tf operations?  shouldn't we be making as much use of simple keras layers as possible?

Reply:
How would you compute a masked softmax with "simple keras layers"? The code I have now is one possible way which I eventually landed on after multiple gradient explosions with my other approaches.

Reply:
huh, ok.  It's a shame that the typical way of dealing with variable sequence length tensors (i.e. for anything text related) can't be used. I'm definitely not familiar with how gradient explosions would be showing up... either you correctly compute gradients which are only on real data you have, or you accidentally compute it on "fake" records.  The latter lead to exploding gradients, you say?

Reply:
Not really. In fact, after my entire weekend of debugging and fixing every possible gradient explosion ever, I still was seeing my loss hit `nan`. And then I gave up and used this instead.

So, loss becomes null when you have `y_pred` as 0/inf/ `y_true_i * lop(y_pred_i)`.
While computing softmax, I had to handle the bounds of `tf.float32` and negative values.
So essentially lots and lots of careful clipping.

Then I realized, I can just pass in a very small value for padded records into the softmax and they would become `0.0` that way and I also get all the stability of tensorflow's softmax with the loss functions.

Reply:
Ok.  I thought that there were layers in keras (for text / sequences) which kinda naturally did this stuff for you...